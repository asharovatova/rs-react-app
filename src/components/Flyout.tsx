import styles from '../app/[locale]/page.module.scss';
import { useDispatch, useSelector } from 'react-redux';
import type { RootState } from '../store/store';
import { clearSelectedItems } from '../store/selectedItemsSlice';
import saveAs from 'file-saver';
import type { CustomPokemon } from '../types/pokemon';
import { useTranslations } from 'next-intl';

export const Flyout = () => {
  const t = useTranslations('flyout');
  const dispatch = useDispatch();
  const selectedPokemons = useSelector(
    (state: RootState) => state.selectedItems.selectedPokemons
  );

  if (selectedPokemons.length === 0) return null;

  const handleDownload = () => {
    const csvContent = [
      ['ID', 'Name', 'Sprite'],
      ...selectedPokemons.map((item: CustomPokemon) => [
        item.id,
        item.name,
        item.sprite,
      ]),
    ]
      .map((row) => row.join(','))
      .join('\n');

    const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' });
    saveAs(blob, `${selectedPokemons.length}_pokemons.csv`);
  };

  return (
    <div className={styles.flyout}>
      <span>
        {selectedPokemons.length} {t('itemsSelected')}
      </span>

      <div className={styles.flexRow}>
        <button onClick={() => dispatch(clearSelectedItems())}>
          {t('unselectAll')}
        </button>

        <button className={styles.buttonDownload} onClick={handleDownload}>
          {t('download')}
        </button>
      </div>
    </div>
  );
};
