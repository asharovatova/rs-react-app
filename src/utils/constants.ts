export const LS_KEY = 'pokemonApp';

export const API_URL = 'https://pokeapi.co/api/v2/pokemon';

export const SPRITE_URL =
  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/';

export const PAGE_LIMIT = 10;

export const NO_IMAGE_AVAILABLE =
  'https://i.pinimg.com/222x/95/d5/cd/95d5cded00f3a3e8a98fb1eed568aa9f.jpg';
