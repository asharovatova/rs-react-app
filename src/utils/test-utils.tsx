import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { MemoryRouter } from 'react-router-dom';
import selectedItemsReducer from '../store/selectedItemsSlice';
import { ThemeProvider } from '../context/ThemeContext';

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    route = '/',
    // theme = 'light',
    // ...renderOptions
  } = {}
) => {
  const store = configureStore({
    reducer: { selectedItems: selectedItemsReducer },
    preloadedState,
  });

  return {
    ...render(
      <ThemeProvider>
        <Provider store={store}>
          <MemoryRouter initialEntries={[route]}>{ui}</MemoryRouter>
        </Provider>
      </ThemeProvider>
    ),
    store,
  };
};
