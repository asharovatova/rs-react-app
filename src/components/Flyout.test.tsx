import { renderWithProviders } from '../utils/test-utils';
import { Flyout } from './Flyout';
import { screen, fireEvent } from '@testing-library/react';
import saveAs from 'file-saver';
import { describe, expect, it, vi } from 'vitest';
import '@testing-library/jest-dom/vitest';

vi.mock('file-saver', () => ({
  default: vi.fn(),
}));

describe('Flyout Component', () => {
  const mockSelectedItems = [
    { id: 1, name: 'Pikachu', sprite: 'pikachu.png' },
    { id: 2, name: 'Charizard', sprite: 'charizard.png' },
  ];

  it('should not render when no items selected', () => {
    renderWithProviders(<Flyout />, {
      preloadedState: { selectedItems: { selectedPokemons: [] } },
    });

    expect(screen.queryByText(/items selected/)).not.toBeInTheDocument();
  });

  it('should render with correct item count', () => {
    renderWithProviders(<Flyout />, {
      preloadedState: {
        selectedItems: { selectedPokemons: mockSelectedItems },
      },
    });

    expect(screen.getByText('2 items selected')).toBeInTheDocument();
  });

  it('should dispatch clearSelectedItems when "Unselect all" clicked', () => {
    const { store } = renderWithProviders(<Flyout />, {
      preloadedState: {
        selectedItems: { selectedPokemons: mockSelectedItems },
      },
    });

    fireEvent.click(screen.getByText('Unselect all'));
    expect(store.getState().selectedItems.selectedPokemons).toEqual([]);
  });

  it('should call saveAs with correct CSV when "Download" clicked', () => {
    renderWithProviders(<Flyout />, {
      preloadedState: {
        selectedItems: { selectedPokemons: mockSelectedItems },
      },
    });

    fireEvent.click(screen.getByText('Download'));

    expect(saveAs).toHaveBeenCalledWith(expect.any(Blob), '2_pokemons.csv');

    [
      ['ID', 'Name', 'Sprite'],
      ['1', 'Pikachu', 'pikachu.png'],
      ['2', 'Charizard', 'charizard.png'],
    ]
      .map((row) => row.join(','))
      .join('\n');

    const blob = vi.mocked(saveAs).mock.calls[0][0] as Blob;
    expect(blob.type).toBe('text/csv;charset=utf-8;');
  });

  it('should match snapshot when visible', () => {
    const { container } = renderWithProviders(<Flyout />, {
      preloadedState: {
        selectedItems: { selectedPokemons: mockSelectedItems },
      },
    });

    expect(container).toMatchSnapshot();
  });
});
