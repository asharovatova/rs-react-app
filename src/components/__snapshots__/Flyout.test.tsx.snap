// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flyout Component > should match snapshot when visible 1`] = `
<div>
  <div
    class="_flyout_efb89e"
  >
    <span>
      2
       items selected
    </span>
    <div
      class="_flexRow_efb89e"
    >
      <button>
        UnselectÂ all
      </button>
      <button
        class="_buttonDownload_efb89e"
      >
        Download
      </button>
    </div>
  </div>
</div>
`;
